### PR TITLE
Run Flow API: Add agentless execution for flow runs

### PR DESCRIPTION
**Pull request by madkinsz**
A small docs blurb is at https://deploy-preview-4589--prefect-docs.netlify.app/orchestration/flow-runs/creation.html#cli

## Summary

Sometimes, it's a bit heavy to go through the whole execution model defined by Prefect
- Register the flow
- Create a flow run
- Start an agent
- Agent deploys the flow run to some infrastructure
- Flow run reports to Prefect backend

This is quite different than local flow runs which are just
- `flow.run()` or `prefect run -p my-flow.py`

The agents add a layer to Prefect in which we help you manage the infrastructure your flow is going to be deployed on. However, when debugging or in some deployment stories, you may want to own that part of your flow's execution. Currently, the only recourse is to use local flow runs in which you lose features that are only supported when backed by the API and inspection of your runs in the UI.

To solve this, I've introduced a new concept of "agentless flow run execution". Here, you take full ownership of the infrastructure your flow runs on. If you want the flow to run in a container, you must set up the container yourself then call this command. This also prevents Prefect from managing the scheduling of the flow run since we cannot spin up the infrastructure with an agent on demand. However, this does allow you to do local runs of flows while still interacting with the full backend API feature set. The execution model here looks like

- Register the flow
- Specify the flow to execute from the CLI, e.g. `prefect run --name "hello-world" --execute`
   -> Creates a flow run for you in the backend with an extra unique label
   -> Creates a subprocess including environment variables from your RunConfig but ignores infrastructure settings
   -> Executes the flow which reports to the API

Some notes of interest:
- If you flow run is scheduled for a time in the future, the subprocess will wait
- If the flow run has a long retry, usually it will release its resources and an agent will pick it up in the future. With this model, the `FlowRunner` will exit and the subprocess will sleep until it should retry
- If the subprocess exits or is killed, the flow run is marked as failed

## Changes

- Adds an `--execute` flag to `prefect run`
- Adds `prefect.backend.execution` to hold execution utilities
- Codifies functions to check for flow/task run retries
- Begins creation of a single function for generate flow run environment variables
- Simplifies `temporary_logger_config`
- Adds utility to convert the config _back_ into environment variables
- Adds handling to `prefect run` for indeterminate final flow run states

## Checklist

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

File: src/prefect/cli/run.py [hunk 1]
```diff
@@ -665,48 +680,73 @@ def run(
...
-        ):
-            level_name = logging.getLevelName(log.level)
-            timestamp = log.timestamp.in_tz(tz="local")
-            echo_with_log_color(
-                log.level, f"└── {timestamp:%H:%M:%S} | {level_name:<7} | {log.message}"
-            )
-
-    except KeyboardInterrupt:
-        quiet_echo("Keyboard interrupt detected!", fg="yellow")
+    # Handle agentless execution
```
Comment (madkinsz) [inline src/prefect/cli/run.py:683]: The diff in here was very weirdly handled by git -- not much has changed. Might be easier to view with indentation changes ignored?

File: src/prefect/utilities/logging.py [hunk 2]
```diff
@@ -294,7 +300,7 @@ def get_logger(name: str = None) -> logging.Logger:
 
 @contextmanager
 def temporary_logger_config(
```
Comment (madkinsz) [inline src/prefect/utilities/logging.py:302]: Vastly simplified this function

Comment (zangell44) [reply src/prefect/utilities/logging.py:302]: niceeee

**Review by zangell44 — COMMENTED** [review]
couple comments, looking great overall!

File: src/prefect/configuration.py [hunk 3]
```diff
@@ -141,6 +141,36 @@ def process_task_defaults(config: Config) -> Config:
     return config
 
 
+def to_environment_variables(
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/configuration.py:144]: I probably missed that -- can do

File: src/prefect/backend/execution.py [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+from prefect.run_configs import RunConfig
+from prefect.utilities.logging import get_logger
+from prefect.utilities.graphql import with_args
+from prefect.configuration import to_environment_variables
+
+
+logger = get_logger("backend.execution")
+
+
+def execute_flow_run_in_subprocess(
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:28]: I'm not sure, but they are used in other modules so it's a bit weird to make them "protected". I do have the disclaimer at the top of the file that this entire module is internal.

File: src/prefect/backend/execution.py [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+
+
+def execute_flow_run(
+    flow_run_id: str,
+    flow: "Flow" = None,
+    runner_cls: Type["prefect.engine.flow_runner.FlowRunner"] = None,
+    **kwargs: Any,
+) -> "FlowRunView":
+    """
+    TODO: This is not used yet and will be hooked up to the CLI in the near future
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:115]: No -- I actually added this TODO. I need to migrate `prefect execute flow-run` to use this

File: src/prefect/backend/execution.py [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+    # Update hardcoded execution variables
+    env.update(
+        {
+            "PREFECT__ENGINE__FLOW_RUNNER__DEFAULT_CLASS": "prefect.engine.cloud.CloudFlowRunner",
+            "PREFECT__ENGINE__TASK_RUNNER__DEFAULT_CLASS": "prefect.engine.cloud.CloudTaskRunner",
+        }
+    )
+
+    # Filter out `None` values
+    return {k: v for k, v in env.items() if v is not None}
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:290]: `subprocess` will throw an error `TypeError: expected str, bytes or os.PathLike object, not NoneType`

File: src/prefect/backend/execution.py [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+
+    # Set up the environment
+    env = generate_flow_run_environ(
+        flow_run_id=flow_run_id,
+        flow_id=flow_run.flow_id,
+        run_api_key=run_api_key,
+        run_config=flow_run.run_config,
+    )
+
+    if flow_run.state.is_running() or flow_run.state.is_submitted():
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:61]: Hm.. you may be right. This is a guard against double execution. All other cases are handled elsewhere. I'll think about this some more.

File: src/prefect/backend/execution.py [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+        }
+    )
+    task_runs = result.get("data", {}).get("task_run")
+    if task_runs is None:
+        raise ValueError(f"Unexpected result while querying for task runs: {result}")
+    elif not task_runs:
+        return None  # No scheduled task runs
+
+    task_runs = sorted(
+        task_runs, key=lambda task_run: task_run.state_start_time, reverse=True
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:376]: 🤦 I didn't have the reverse then I wrote tests and thought it was wrong so I added it. I think that you're right and I should change the tests for this. I can also try `min`

File: src/prefect/backend/execution.py [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+        # current context
+        **secrets,
+        **run_kwargs["context"].get("secrets", {}),
+    }
+    # Update some default run kwargs with flow settings
+    run_kwargs.setdefault("executor", flow.executor)
+
+    # Execute the flow, this call will block until exit
+    logger.info(
+        f"Beginning execution of flow run {flow_run.name!r} from {flow_metadata.name!r} "
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:188]:
```suggestion
        f"Beginning execution of flow run {flow_run.name!r} from flow {flow_metadata.name!r} "
```

File: src/prefect/backend/execution.py [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+    Returns:
+        A `FlowRunView` instance with information about the state of the flow run and its
+        task runs
+    """
+    logger.debug(f"Querying for flow run {flow_run_id!r}")
+
+    # Get the `FlowRunner` class type
+    # TODO: Respect a config option for this class so it can be overridden by env var,
+    #       create a separate config argument for flow runs executed with the backend
+    runner_cls = runner_cls or prefect.engine.cloud.flow_runner.CloudFlowRunner
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:143]: I think I'll address this when this function is actually used

File: src/prefect/backend/execution.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/backend/execution.py:28]: I could change the whole module to `_execution.py`

**Comment by madkinsz** [comment]
@zangell44 I've addressed everything I think!

**Review by zangell44 — DISMISSED** [review]

File: src/prefect/backend/execution.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zangell44) [reply src/prefect/backend/execution.py:28]: nah that's fine I think we're clear enough for right now

**Review by zangell44 — APPROVED** [review]

**Pull request merged**